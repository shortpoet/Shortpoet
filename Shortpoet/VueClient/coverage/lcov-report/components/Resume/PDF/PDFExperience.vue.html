
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/Resume/PDF/PDFExperience.vue</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">components/Resume/PDF</a> PDFExperience.vue</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>2/2</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>4/4</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>2/2</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>2/2</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line high'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">&lt;template&gt;
  &lt;section :class="sectionClass" id="experience"&gt;
    &lt;div class="my-auto"&gt;
      &lt;h4 style="font-family: 'Saira Extra Condensed';"&gt;Experience&lt;/h4&gt;
      &lt;div class="d-flex flex-column justify-content-between" v-for="type in experiences" :key="type.type"&gt;
        &lt;!-- &lt;h4 class="pdf-subheading mt-2 mb-0"&gt;{{ type.type }} Experience&lt;/h4&gt; --&gt;
        &lt;div v-for="(job, ji) in type.jobs" :key="ji"&gt;
          &lt;div class="d-flex flex-column justify-content-between"&gt;
            &lt;div class="d-flex flex-row"&gt;
              &lt;div class="mr-auto"&gt;
                &lt;h5 id="job-position" style="font-family: 'Saira Extra Condensed';" class="mb-0"&gt;{{ job.position }}&lt;/h5&gt;
              &lt;/div&gt;
              &lt;div :class="textRightClass"&gt;
                &lt;h6 style="font-family: 'Saira Extra Condensed';" class="pdf-experience-dates text-primary"&gt;{{ job.startDate }} - {{ job.endDate }}&lt;/h6&gt;
              &lt;/div&gt;
            &lt;/div&gt;
            &lt;div class="d-flex flex-column"&gt;
              &lt;h6 style="font-family: 'Saira Extra Condensed';" class="pdf-company"&gt;{{ job.company }}&lt;/h6&gt;
              &lt;div
                v-for="(text, di) in job.description.split('\\n\\r')"
                :key="di"
                class="mb-2"
              &gt;
                &lt;p style="font-family: 'Open Sans';" v-html="text" :class="'job-' + (ji + 1) + '-description-' + (di + 1)"&gt;&lt;/p&gt;
              &lt;/div&gt;
&nbsp;
            &lt;/div&gt;
          &lt;/div&gt;
        &lt;/div&gt;
      &lt;/div&gt;
    &lt;/div&gt;
  &lt;/section&gt;
&lt;/template&gt;
&lt;script&gt;
export default {
  name: 'PDFExperience',
  props: {
    experiences: {
      type: Array
    },
    renderPDF: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    sectionClass () {
      return this.renderPDF ?
        'resume-section-render ml-3 mt-2 mb-3'
        :
        'resume-section ml-3 mt-2 mb-3'
    },
    textRightClass () {
      return this.renderPDF ?
        'resume-date text-md-right mr-1'
        :
        'resume-date text-md-right mr-4'
    }
  }
}
&lt;/script&gt;
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Sun May 03 2020 17:20:17 GMT-0500 (Central Daylight Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    