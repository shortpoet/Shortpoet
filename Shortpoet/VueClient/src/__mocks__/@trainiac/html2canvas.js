
const html2canvas = jest.genMockFromModule('@trainiac/html2canvas')

console.log('hello from html2canvas mock')

module.exports = html2canvas