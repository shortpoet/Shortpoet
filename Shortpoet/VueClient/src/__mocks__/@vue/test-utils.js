// export default {
//   mount: jest.fn(() => {
//     console.log("mount mock")
//   })
// }

const test = jest.genMockFromModule('@vue/test-utils')
const component = {}
const mountOptions = {
  propsData: {},
  localVue: {},
mocks: {},
store: {},
context: {}, // only for functional components
router: {},
computed: {},
stubs: {},
slots: {},
scopedSlots: {},
attrs: {},
attachToDocument: false,
attachTo: null, //HTML Element or String
listeners: {},
parentComponent: {},
provide: {}
}

test.mount = jest.fn(() => {
  return {
    component,
    mountOptions
  }
})

module.exports = test