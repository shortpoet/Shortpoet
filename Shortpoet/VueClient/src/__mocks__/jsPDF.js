
const jsPDF = jest.genMockFromModule('jsPDF')

console.log('hello from jsPDF mock')

module.exports = jsPDF