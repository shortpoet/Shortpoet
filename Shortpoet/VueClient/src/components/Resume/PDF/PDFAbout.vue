<template>
  <section class="resume-section" id="about">

    <!-- basic info row -->
    <div class="d-flex flex-row justify-content-between align-items-center mt-1 ml-3" id="basic-info">

      <!-- photo column -->
      <div class="d-flex flex-column align-items-center" id="photo">
        <img class="img-fluid img-profile rounded m-1 mb-2" :src="pic" alt="profile pic" :style="{height: '8rem', width: '7rem'}">
      </div>

      <!-- name contact column -->
      <div class="d-flex flex-column mt-1 mt-md-2 ml-2" id="name-contact">
        <ul class="list-inline">
          <li>
            <h2 style="font-family: 'Saira Extra Condensed';" class="mb-0">{{ name }}
              <span id="pdf-surname" style="font-family: 'Saira Extra Condensed';" class="text-primary">{{ surname }}</span>
            </h2>
          </li>
          <li>
            <span class="not-name">
              <a style="font-family: 'Open Sans';" :href="'mailto:' + email">
                {{ email }}
              </a>
            </span>
          </li>
          <li>
            <span id="pdf-address" style="font-family: 'Saira Extra Condensed';" class="not-name">
              {{ address }}
            </span>
          </li>
          <li>
            <span id="pdf-flags" class="not-name">{{ flags.replace(/,/g, '') }}</span>
          </li>
          <li>
            <span id="pdf-visas" style="font-family: 'Saira Extra Condensed';" class="not-name">Valid Work Visas: {{ visas }}</span>
          </li>
        </ul>
      </div>

      <!-- socials column -->
      <div class="d-flex flex-column ml-auto mr-1 mr-md-3" id="socials">
        <PDFSocials :renderPDF="renderPDF" />
      </div>

    </div>


  </section>
</template>

<script>
// import PDFAction from '@/components/Utils/PDFAction'
import PDFSocials from '@/components/Resume/PDF/PDFSocials'

export default {
    name: 'PDFAbout',
    components: {
      // PDFAction,
      PDFSocials
    },
    props: {
      name: {
        type: String
      },
      surname: {
        type: String
      },
      email: {
        type: String
      },
      address: {
        type: String
      },
      visas: {
        type: String
      },
      flags: {
        type: String
      },
      renderPDF: {
        type: Boolean,
        default: false
      }
    },
    data () {
      return {
        pic: require('@/assets/github_profile_pic.png'),
        socials: [
          {
            social: 'github',
            url: 'https://github.com/shortpoet',
            transform: 'shrink-5'
          },
          {
            social: 'linkedin',
            url: 'https://www.linkedin.com/in/carlos-soriano-49aaa97/',
            transform: 'shrink-6 right-1'
          },
          {
            social: 'instagram',
            url: 'https://www.instagram.com/shortpoet/',
            transform: 'shrink-5 right-1'
          },
          {
            social: 'twitter',
            url: 'https://twitter.com/shortpoet3',
            transform: 'shrink-6'
          },
          {
            social: 'globe',
            url: 'https://shortpoet.com',
            transform: 'shrink-6 right-.25'
          }
        ]
      }
    },
    computed: {
    },
    mounted () {
    }
}
</script>
<style>
.social-item {
  font-size: .55rem;
}
</style>
