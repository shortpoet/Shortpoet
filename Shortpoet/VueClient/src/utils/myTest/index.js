// address
import createWrapper from './test.utils.createWrapper'
import factory from './test.utils.factory'
import factoryShallow from './test.utils.factoryShallow'
export default {
  createWrapper,
  factory,
  factoryShallow
}
