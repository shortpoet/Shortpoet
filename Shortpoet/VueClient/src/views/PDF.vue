<template>
  <div class="" v-if="getResumeLoaded" id="resume-anchor">
    <div class="p-10">
      <PDFAbout
        :name="getResume.name"
        :surname="getResume.surname"
        :email="getResume.email"
        :aboutMe="getResume.aboutMe"
        :address="getResume.address"
      />
      <PDFExperience
        :experiences="getResume.experiences"      
      />
      <PDFEducation
        :educations="getResume.educations"
      />
      <PDFSkills
        :skills="getResume.skills"
      />
      <PDFInterests
        :awards="getResume.spokenLanguages"
      />
      <PDFAwards
        :interests="getResume.interests"
      />
    </div>
  </div>
</template>

<script>
import PDFAbout from '@/components/Resume/PDF/PDFAbout'
import PDFExperience from '@/components/Resume/PDF/PDFExperience'
import PDFEducation from '@/components/Resume/PDF/PDFEducation'
import PDFSkills from '@/components/Resume/PDF/PDFSkills'
import PDFInterests from '@/components/Resume/PDF/PDFInterests'
import PDFAwards from '@/components/Resume/PDF/PDFAwards'
import { mapGetters, mapActions } from 'vuex'

// const yaml = require('js-yaml')
// import * as fs from 'fs'

export default {
  name: 'Resume',
  components: {
    PDFAbout ,
    PDFExperience,
    PDFEducation,
    PDFSkills,
    PDFInterests,
    PDFAwards
  },
  data () {
    return {
    }
  },
  computed: {
    ...mapGetters('resume', ['getResume', 'getResumeLoaded']),
    resume () {
      return this.getResume.skills.forEach(skill => {
        console.log(skill)
        skill.details = skill.details.split(',')
        console.log(skill)
      })
    }
    
  },
  methods: {
    ...mapActions('resume', ['loadResume']),
  },
  mounted () {
    const $ = this.jquery
    this.$nextTick(() => {
      // possibly move this into utils?  check if global vue jquery $ variable is still needed in that case.  or at all for that matter.
      // our custom jQuery code goes here
      $('a.js-scroll-trigger[href*="#"]:not([href="#"])').click(function () {
        if (location.pathname.replace(/^\//, '') === this.pathname.replace(/^\//, '') && location.hostname === this.hostname) {
          var target = $(this.hash)
          target = target.length ? target : $('[name=' + this.hash.slice(1) + ']')
          if (target.length) {
            $('html, body').animate({
              scrollTop: (target.offset().top)
            }, 1000, 'easeInOutExpo')
            return false
          }
        }
      })
      // Closes responsive menu when a scroll trigger link is clicked
      $('.js-scroll-trigger').click(function () {
        $('.navbar-collapse').collapse('hide')
      })
      // Activate scrollspy to add active class to navbar items on scroll
      $('body').scrollspy({
        target: '#sideNav'
      })
      this.loadResume()
    })
  }
}
</script>

<style lang="scss">

@import url('https://fonts.googleapis.com/css?family=Saira+Extra+Condensed:100,200,300,400,500,600,700,800,900');
@import url('https://fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i,800,800i');
@import './../assets/scss/pdf.scss';

</style>
