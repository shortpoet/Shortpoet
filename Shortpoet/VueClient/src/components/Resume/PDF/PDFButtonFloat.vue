<template>
  <portal :to="target">
    <!-- 
      * previously had this selector to unify the components - split after styling diverged
      * v-if="href" 
      * v-if="pdfTarget"
    </div> -->
    <div type="input" :class="classObject" @click="showModal">
      <font-awesome-layers
        class="button-float-icon-layer fa-lg"
      >
        <font-awesome-icon class="button-float-icon-circle" size="2x" icon="circle" />
        <font-awesome-icon
          class="button-float-icon"
          size="2x"
          :transform="_icon.transform"
          :icon="_icon.icon"
        ></font-awesome-icon>
      </font-awesome-layers>
    </div>
    <div class="modal-slot">
      <PDFModal
        v-show="isModalVisible"
        @close="closeModal"
        @to-pdf="toPDF"
        @to-page="toPage"
        @to-canvas="toCanvas"
      >
        <!-- <template v-slot:header>
          <h1>Test Header</h1>
        </template>
        <template v-slot:body>
          <h1>Test Body</h1>
        </template>
        <template v-slot:footer>
          <h1>Test Footer</h1>
        </template> -->
      </PDFModal>
    </div>
  </portal>
</template>

<script>
import PDFModal from '@/components/Resume/PDF/PDFModal'
import { colorLog } from '@/utils/colorLog'
// import { log } from '@/utils/colorLog'
// log('blue', 'test')
import jsPDF from 'jspdf'
// using fork for now to solve this issue
// the changes i made with the icons as the deployed version still works
// https://github.com/niklasvh/html2canvas/issues/1868#issuecomment-599217709
import html2canvas from '@trainiac/html2canvas'
// import html2canvas from 'html2canvas'
const FontFaceObserver = require('fontfaceobserver');

export default {
  name: 'PDFButtonFloat',
  components: {
    PDFModal
  },
  props: {
    target: {
      type: String
    },
    icon: {
      type: String
    },
    pdfTarget: {
      type: String
    }
    // handler: {
    //   type: Function,
    //   required: false
    // }
  },
  data () {
    return {
      iconMap: {
        pdf: {
          icon: 'file-pdf',
          transform: 'shrink-5 right-2.3'
        },
        save: {
          icon: 'save',
          transform: 'shrink-5 right-1'
        }
      },
      isModalVisible: false,
    }
  },
  computed: {
    mobile () {
      return window.innerWidth < 768
    },
    classObject () {
      return {
        'button-float': true,
        'mobile': this.mobile
      }
    },
    _icon () {
      return this.iconMap[`${this.icon}`]
    }
  },
  methods: {
    showModal() {
      this.isModalVisible = true
    },
    closeModal() {
      this.isModalVisible = false
    },
    /* istanbul ignore next */
    toCanvas () {
      html2canvas(document.getElementById(this.pdfTarget), {
          allowTaint: true,
        }).then(canvas => {
          document.body.appendChild(canvas);
        })
    },
    async checkFonts () {
      var fontA = new FontFaceObserver('Open Sans')
      var fontB = new FontFaceObserver('Saira Extra Condensed')
      // must return otherwise load order is not correct
      return Promise.all([fontA.load(), fontB.load()]).then(function () {
        console.log('Family A & B have loaded')
      })
    },
    async getCanvas (options) {
      const vm = this
      const target = document.getElementById(vm.pdfTarget)
      return html2canvas(target, {
        ...options
      })
    },
    // close modal first
    // [Vue warn]: Method "isModalVisible" has type "boolean" in the component definition. Did you reference the function correctly?      this.closeModal()
    // [Vue warn]: Method "isModalVisible" has already been defined as a data property.
    // vm.isModalVisible = false
    // vm.closeModal()
    // both threw those warning but only in test
    // realized that the click event on the modal buttons 
    // bubbled up to parent backdrop div that closes on click
    // which is why the modal tests had that double event that had me wondering
    // seems then no need to call close from here which makes sense as far as separation of concerns i guess
    //
    // setting width and height cuts it off at the "one page mark"
    // width: 810,
    // height: 1100,      
    async toPDF() {
      const vm = this
      colorLog('toPDF from Button Float', 'blue')
      await vm.checkFonts()
      colorLog('fonts have been checked', 'violet')
      setTimeout(async () => {
        const canvas = await vm.getCanvas({
          scale: 5,
          useCORS: true,
          allowTaint: true,
        })
        const image = canvas.toDataURL('image/jpeg', 1.0);
        const doc = new jsPDF('p', 'mm', 'a4');
        const pageWidth = doc.internal.pageSize.getWidth();
        const pageHeight = doc.internal.pageSize.getHeight();
        const widthRatio = pageWidth / canvas.width;
        const heightRatio = pageHeight / canvas.height;
        const ratio = widthRatio > heightRatio ? heightRatio : widthRatio;
        const canvasWidth = canvas.width * ratio;
        const canvasHeight = canvas.height * ratio;
        const marginX = 0 //(pageWidth - canvasWidth) / 2;
        const marginY = 0 //(pageHeight - canvasHeight) / 2;
        doc.addImage(image, 'JPEG', marginX, marginY, canvasWidth, canvasHeight, null, 'SLOW');
        doc.save(`Carlos_Soriano_${Date.now()}.pdf`);
      }, 250);
    },
    // https://stackoverflow.com/questions/24069124/how-to-save-a-image-in-multiple-pages-of-pdf-using-jspdf
    // https://stackoverflow.com/questions/19272933/jspdf-multi-page-pdf-with-html-renderer/34934497#34934497
    // insterstingly this used to work even when pdf was not reassigned to paginated
    // eg 
    // vm.paginate(target, canvas, pdf);
    // paginated.save(`Carlos_Soriano_${Date.now()}.pdf`);
    // vm.$emit('toPDF', false);
    async toPage () {
      const vm = this
      vm.$emit('to-pdf', true)
      const target = document.getElementById(vm.pdfTarget)
      await vm.checkFonts()
      colorLog('fonts have been checked', 'violet')
      setTimeout(async ()=>{
        const canvas = await vm.getCanvas({
          width: 810,
          scale: 5,
          useCORS: true,
          allowTaint: true
        })
        var pdf = new jsPDF('p', 'pt', 'a4');
        const paginated = vm.paginate(target, canvas, pdf);
        paginated.save(`Carlos_Soriano_${Date.now()}.pdf`);
        vm.$emit('toPDF', false);
      }, 250);
    },
    paginate (target, canvas, pdf) {
      const imgData = canvas.toDataURL('image/png')
      const imgWidth = 595
      const pageHeight = 842
      let imgHeight = canvas.height * imgWidth / canvas.width;
      let heightLeft = imgHeight
      pdf.addImage(imgData, 'PNG', 0, 0, imgWidth, imgHeight, null, 'SLOW')
      heightLeft -= pageHeight
      let secondHalf = heightLeft - imgHeight
      pdf.addPage()
      pdf.addImage(imgData, 'PNG', 0, secondHalf, imgWidth, imgHeight, null, 'SLOW')
      return pdf
    }
  },
  mounted () {
  }
}
</script>

<style lang="scss">

</style>