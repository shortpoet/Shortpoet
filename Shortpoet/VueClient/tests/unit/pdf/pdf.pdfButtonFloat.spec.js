// TODO organize rest of tests' imports like this

// test utils
import * as myTest from '@/utils/myTest'
const { createWrapper } = myTest.default
import { createLocalVue } from '@vue/test-utils'

// utils
import { cloneDeep } from 'lodash'
import Vue from 'vue'

// plugins
import PortalVue from 'portal-vue'
import { FontAwesomeIcon, FontAwesomeLayers } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faCircle, faFilePdf, faSave, faTimes, faRocket} from '@fortawesome/free-solid-svg-icons'

// SUT
import PDFButtonFloat from '@/components/Resume/PDF/PDFButtonFloat'
import PDF from '@/views/PDF'
import hardResume from '@/assets/resume.js'

// mocks
import jsPDF from '@/__mocks__/jspdf'
// similar to mocking @vue/testutils
// folders had to be nested just like source module for mock to work
import html2canvas from '@/__mocks__/@trainiac/html2canvas'
import fontfaceobserver from '@/__mocks__/fontfaceobserver'

let wrapper

describe('start.pdfButtonFloat', () => {

  const component = PDFButtonFloat

  let methods = component.methods

  const LocalVue = createLocalVue()
  LocalVue.use(PortalVue)

  let propsData

  let computed

  let stubs

  let mocks

  let ignoredElements

  let resumeStoreOptions

  let mountOptions

  let expected

  const events =  ['click', 'touchstart', 'touchcancel', 'touchmove', 'touchend']

  beforeEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
    expected = ``

    library.add(
      faCircle, faFilePdf, faSave, faTimes, faRocket
    )

    stubs = {
      FontAwesomeIcon,
      FontAwesomeLayers
    }
  
    propsData = {
      target: 'save-button-float',
      icon: 'save',
      pdfTarget: 'pdf-anchor'
    }
  
    computed = {}


    ignoredElements = [
      // 'portal'
    ]

    mocks = {
    }
    
    mountOptions = {
      propsData: propsData,
      stubs: stubs
    }

    // mountOptions.attachToDocument = true

    const getResumeLoaded = true

    resumeStoreOptions = { 
      getters: { getResumeLoaded: jest.fn(() => getResumeLoaded), getResume: jest.fn(() => hardResume) }, 
      mutations: { 'SET_RESUME_RAW': jest.fn() } 
    }

    // wrapper = createWrapper(component, mountOptions, resumeStoreOptions)

  })

  describe('pdf.pdfButtonFloat.toPDF', () => {

    it('modal should already be closed by bubbled emit event and by default', () => {
      
      // methods.toPDF()
      // expect(wrapper.vm.isModalVisible).toBe(false)
    })

    it('canvas context should match snapshot', async () => {
      /**
       * In order to see which functions and properties were used for the test, you can use `__getEvents`
       * to gather this information.
       */
      // const events = ctx.__getEvents()
      
      // expect(events).toMatchSnapshot()
      // const canvas = await methods.toPDF()
      // console.log(canvas)
      // expect(html2canvas).toHaveBeenCalled()
    })
  })

  describe('pdf.pdfButtonFloat.showModal', () => {
    it('should show modal by changing isModalVisible to true', () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions)
      expect(wrapper.vm.isModalVisible).toBe(false)
      // 1st error being thrown seemed to be from test thinking this.isModalVisible was a function assignment
      // fixed by changing to .call()
      // then error bec no this context in test to added wrapper.vm
      // https://medium.com/@lachlanmiller_52885/data-and-interaction-testing-in-vue-e7914a9179d7
      methods.showModal.call(wrapper.vm)
      expect(wrapper.vm.isModalVisible).toBe(true)
    })
  })

  describe('pdf.pdfButtonFloat.closeModal', () => {
    it('should close modal by changing isModalVisible to false', async () => {
      const data = () => {
        return {
          isModalVisible: true
        }
      }
      mountOptions = {
        propsData: propsData,
        data: data
      }
  
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions)

      expect(wrapper.vm.isModalVisible).toBe(true)
      methods.closeModal.call(wrapper.vm)
      expect(wrapper.vm.isModalVisible).toBe(false)
    })
  })

  describe('pdf.pdfButtonFloat.checkFonts', () => {
    it('should call fontfaceobserver', async () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions)
      await methods.checkFonts.call(wrapper.vm)
      expect(fontfaceobserver).toHaveBeenCalled()
    })
  })

  describe('pdf.pdfButtonFloat.getCanvas', () => {
    it('should call html2canvas with options', async () => {

      mocks = {
        dispatch: jest.fn()
      }

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true
      }

      // had to create wrapper from parent component because using vue portal
      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions)

      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true,
      }

      // this wasy doesn't load the context needed for method to run correctly
      // await methods.getCanvas(options)

      await wrapper.find(PDFButtonFloat).vm.getCanvas(options)
      
      const target = document.getElementById(wrapper.find(PDFButtonFloat).vm.pdfTarget)
      
      expect(html2canvas).toHaveBeenCalledWith(target, options)
    
    })
  })
  describe('pdf.pdfButtonFloat.snapshot', () => {
    it('should match snapshot', () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions)

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

})
