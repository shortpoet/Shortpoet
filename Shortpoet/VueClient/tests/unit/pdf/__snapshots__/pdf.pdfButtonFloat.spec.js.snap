// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start.pdfButtonFloat pdf.pdfButtonFloat.snapshot should match snapshot 1`] = `<div class="v-portal" style="display: none;"></div>`;
